File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.java
import java.io.*;
import java.util.*;

/**
 * 가장 심플한 아이디어 : 벽 하나를 임의로 제거한 다음, 최단 거리를 구한다. -> O(n^4) 불가능
 * 상태 유지 아이디어 : 큐에 벽을 부수고 이동하는지 여부를 같이 저장한다. 
 * -> 특정 지점에서 최단거리 기록할 때 벽 부순 유무 저장 
 * -> 복잡
 *  d[n][m] = (0,0) -> (벽) + (벽) -> (n, m)
 * 1. 벽을 부수지 않는다고 가정하고 모든 지점까지의 최단 거리를 구한다. O(n^2)
 * 2. 벽에서 출발했을 때, 최단 거리 구하기
 */

public class Main {

    static char[][] board = new char[1002][1002];
    static int[][] dist = new int[1002][1002];
    static int n;
    static int m;
    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] parts = br.readLine().split(" ");
        n = Integer.parseInt(parts[0]);
        m = Integer.parseInt(parts[1]);

        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < m; j++) {
                board[i][j] = line.charAt(j);
                dist[i][j] = -1;
            }
        }

        Queue<Pair> queue = new ArrayDeque<>();
        Queue<Pair> wallQueue = new ArrayDeque<>();
        queue.offer(new Pair(0, 0));
        dist[0][0] = 1;

        while (!queue.isEmpty()) {
            var cur = queue.poll();

            for (int i = 0; i < 4; i++) {
                int nx = cur.x + dx[i];
                int ny = cur.y + dy[i];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if (dist[nx][ny] != -1) continue;
                dist[nx][ny] = dist[cur.x][cur.y] + 1;

                if (board[nx][ny] == '1') wallQueue.offer(new Pair(nx, ny));
                else queue.offer(new Pair(nx, ny));
            }
        }

        while (!wallQueue.isEmpty()) {
            var cur = wallQueue.poll();

            for (int i = 0; i < 4; i++) {
                int nx = cur.x + dx[i];
                int ny = cur.y + dy[i];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if (board[nx][ny] == '1') continue;
                if (dist[nx][ny] != -1 && dist[nx][ny] <= dist[cur.x][cur.y] + 1) continue;
                dist[nx][ny] = dist[cur.x][cur.y] + 1;
                wallQueue.offer(new Pair(nx, ny));
            }
        }

        System.out.print(dist[n - 1][m - 1]);
    }

    static class Pair {
        int x;
        int y;

        public Pair(int xx, int yy) {
            x = xx;
            y = yy;
        }
    }
}

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.kt
import java.util.ArrayDeque
import java.util.Queue
import kotlin.math.min

/**
 * 문제: 출발점에서 도착지까지 최단 이동 거리 구하기
 * - 조건: 벽을 1회 부술 수 있다.
 * - n, m <= 1000
 * 각 벽을 부수는 경우에 대해서는 시간이 부족함
 * 접근: 상태를 가지고 이동해보자.
 * 벽을 부셨는지, 안부셨는지를 기록하면서 이동한다.
 * 상수배 연산이 늘어나므로, 시간 복잡도상 괜찮음
 */

val dx = arrayOf(1, 0, -1, 0)
val dy = arrayOf(0, 1, 0, -1)
const val INF = 0x3f3f3f3f
const val MAX_BREAK_COUNT = 1

fun main() = with(System.`in`.bufferedReader()) {
    val (rows, cols) = readLine().split(" ").map { it.toInt() }

    val road = Array(rows) { IntArray(cols) }
    for (i in 0 until rows) {
        readLine().toCharArray()
            .map { it - '0' }
            .forEachIndexed { idx, x -> road[i][idx] = x }
    }

    // 상태 저장
    val dist = Array(rows) { Array(cols) { IntArray(2) { INF } } }
    val queue: Queue<Triple<Int, Int, Int>> = ArrayDeque()

    dist[0][0][0] = 1
    queue.offer(Triple(0, 0, 0))

    while (queue.isNotEmpty()) {
        val (cx, cy, count) = queue.poll()

        for (i in 0 until 4) {
            val nx = cx + dx[i]
            val ny = cy + dy[i]

            if (nx < 0 || ny < 0 || nx >= rows || ny >= cols) continue;

            if (count < MAX_BREAK_COUNT) {
                // 벽 뚫는 경우
                if (road[nx][ny] == 1 && dist[nx][ny][1] == INF) {
                    dist[nx][ny][1] = dist[cx][cy][0] + 1
                    queue.offer(Triple(nx, ny, count + 1))
                }

                // 벽 안뚫는 경우
                if (road[nx][ny] == 0 && dist[nx][ny][0] == INF) {
                    dist[nx][ny][0] = dist[cx][cy][0] + 1
                    queue.offer(Triple(nx, ny, count))
                }

            } else {
                // 더이상 벽 못뚫음
                if (road[nx][ny] == 0 && dist[nx][ny][1] == INF) {
                    dist[nx][ny][1] = dist[cx][cy][1] + 1
                    queue.offer(Triple(nx, ny, count))
                }
            }
        }
    }

    val answer = min(dist[rows - 1][cols - 1][0], dist[rows - 1][cols - 1][1])
    print(if (answer == INF) -1 else answer)
}
